// router/router.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _logger = require('logger');

var _logger2 = _interopRequireDefault(_logger);

var _preData = require('preData');

var _preData2 = _interopRequireDefault(_preData);

var React = require('react/addons');
var rRouter = require('react-router');
var Route = rRouter.Route;
var NotFoundRoute = rRouter.NotFoundRoute;

var logger = _logger2['default'].getLogger('router');

function escapeHtml(text) {
	var map = {
		'&': '&amp;',
		'<': '&lt;',
		'>': '&gt;',
		'"': '&quot;',
		"'": '&#039;'
	};
	return text.replace(/[&<>"']/g, function (m) {
		return map[m];
	});
}

var render = function render(layout, params) {
	for (var _i in params) {
		var i = params[_i];
		layout = layout.split("%" + _i + "%").join(i);
		layout = layout.split("#" + _i + "#").join(escapeHtml(i));
	}
	return layout.replace(/(\n\s+)/gi, "").replace(/(\n)/gi, "");
};

var Router = (function () {
	function Router() {
		_classCallCheck(this, Router);

		this.routes = [];
	}

	_createClass(Router, [{
		key: 'add',
		value: function add(path, handler, status, title, express) {
			this.routes.push({
				path: path || '/',
				handler: handler || null,
				status: status || null,
				title: title || ':)',
				express: express || function (ret) {
					ret.end();
				}
			});
		}
	}, {
		key: 'rxReact',
		value: function rxReact(Layout) {
			var reactElements = [Route, { handler: Layout, location: "history" }];
			this.routes.forEach(function (e) {
				reactElements.push(React.createElement(Route, { path: e.path, handler: e.handler }));
			});
			var routes = React.createElement.apply(React, reactElements);
			return routes;
		}
	}, {
		key: 'exec',
		value: function exec(app, template, Layout, config) {
			var browserify = require('browserify' + '-middleware');
			app.use('/api/session', function (request, response, next) {
				response.json(request.session);
			});
			app.use(config.browser.bundle || '/bundle.js', browserify(require('path').join('./application/', config.react.browser), {
				transform: ["babelify", "require-globify"],
				stage: 0,
				minify: !config.debug,
				cache: !config.debug
			}));
			var routes = this.rxReact(Layout);
			var preData = new _preData2['default'](app, this.routes);
			app.use(function (request, response, next) {
				if (!request.__ret) {
					logger.warn(request.path);
					response.status(404);
					response.end('Not found 404');
					return;
				}

				var _ref = request.__ret || {};

				var req = _ref.req;
				var res = _ref.res;
				var status = _ref.status;

				var relStatus = request.session.account.status;
				if (status < relStatus) {
					logger.warn(request.path);
					response.status(403);
					response.end('Error 403');
					return;
				}
				logger.info(request.path);
				logger.info('==============');
				logger.info(res.data);
				logger.info('==============');
				var dataId = preData.create(res.data);
				var url = request.path + '?dataId=' + dataId;
				rRouter.run(routes, url, function (Root) {
					var content = React.renderToString(React.createElement(Root, null));
					response.end(render(template, {
						title: res.title || '',
						preData: !request.query.script ? JSON.stringify(preData.remove(dataId)) : '{}',
						script: !request.query.script ? '' : '-tmp',
						style: !request.query.style ? '' : '-tmp',
						styleurl: config.style.bundle || '',
						scripturl: config.browser.bundle || '',
						session: !request.session ? '{}' : JSON.stringify(request.session),
						'yield': content || ''
					}));
				});
			});
			return app;
		}
	}]);

	return Router;
})();

module.exports = Router;
