// plugins/server.js
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _express = require('express');

var _express2 = _interopRequireDefault(_express);

var _configs = require('configs');

var _configs2 = _interopRequireDefault(_configs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var list = (0, _configs2['default'])('plugins');

var router = _express2['default'].Router();

var plugins = {};

var Session = require('./session.js');

var Plugin = (function () {
	function Plugin(parent) {
		var _this = this;

		_classCallCheck(this, Plugin);

		this.__parent = parent;
		this.sync = !!this.__parent.sync;

		var _loop = function (_name) {
			var body = _this.__parent[_name];
			if (typeof body == 'function') {
				_this[_name] = function (options, session) {
					return _this.exec(_name, options, session);
				};
			}
		};

		for (var _name in this.__parent) {
			_loop(_name);
		};
		this.getPlugin = function (name) {
			return plugins[name];
		};
	}

	_createClass(Plugin, [{
		key: 'exec',
		value: function exec() {
			var name = arguments.length <= 0 || arguments[0] === undefined ? null : arguments[0];

			var _this2 = this;

			var options = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];
			var session = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];

			if (this.sync) {
				return this.__parent[name].call(this.__parent, options);
			} else {
				if (session && !session['__isSession']) {
					session = new Session(session);
				}
				var promise = new Promise(function (resolve, reject) {
					if (_this2.__parent[name] != undefined) {
						_this2.__parent[name].call(_this2, options, session, resolve, reject);
					} else {
						reject('not found method');
					};
				});
				return promise;
			}
		}
	}]);

	return Plugin;
})();

for (var i in list) {
	if (list.hasOwnProperty(i)) {
		var e = list[i];
		var tmpl = require(_path2['default'].resolve('./application/plugins/' + i + '/server/index.js'));
		var rpc = require(_path2['default'].resolve('./application/plugins/' + i + '/client/index.js')).rpc;
		tmpl['name'] = i;
		tmpl['rpc'] = rpc || [];
		var plugin = new Plugin(tmpl);
		plugins[i] = plugin;
	}
}

for (var i in list) {
	if (list.hasOwnProperty(i)) {
		var e = list[i];
		var init = plugins[i].__parent.init || function () {};
		// var app = require('plugins/singleton.js').app;
		init.call(plugins[i], router, e);
	}
}

var Singleton = {
	list: plugins,
	router: router
};

Singleton.__instance = null;

Singleton.__getInstance = function () {
	if (this.__instance === null) {
		this.__instance = Singleton;
	}
	return this.__instance;
};

module.exports = Singleton.__getInstance();
