// plugins/session.es6.js


class Session {
	constructor(tmp) {
		this.__isSession = true;
		var status = 8;
		var id = null;
		var data = {};
		var storage = {};
		var dataSign = null;
		var accSign = null;
		if(typeof tmp == 'object') {
			if(typeof tmp.account == 'object') {
				status = tmp.account.status||status;
				id = tmp.account.id||id;
			}
			if(typeof tmp.data == 'object') {
				data = tmp.data||data;
			}
			if(typeof tmp.storage == 'object') {
				storage = tmp.storage||storage;
			}
			if(typeof tmp.dataSign == 'string') {
				dataSign = tmp.dataSign||dataSign;
			}
			if(typeof tmp.accSign == 'string') {
				accSign = tmp.accSign||accSign;
			}
		}
		this.__data = {
			account: {
				status: status,
				id: id
			},
			dataSign: dataSign,
			accSign: accSign
		};
		this.storage = storage||storage;
		this.data = data||data;
	};
	set id(data) {
		this.__data.account.id = data;
	};
	get id() {
		return this.__data.account.id;
	};
	set status(data) {
		this.__data.account.status = data;
	};
	get status() {
		return this.__data.account.status;
	};
	sign(data) {
		return signature.sign(data);
	};
	export() {
		var ret = {};
		var storage = {};
		var data = {};
		if(typeof this.data == 'object') {
			data = this.data||data;
		};
		if(typeof this.storage == 'object') {
			storage = this.storage||storage;
		};
		ret['account'] = this.__data.account;
		ret['storage'] = storage;
		ret['data'] = data;
		return ret;
	};
	exportPacked() {
		var encode = function(data) {
			return new Buffer(data).toString('base64');
		};
		var decode = function(data) {
			return new Buffer(data, 'base64').toString('ascii');
		};
		var ret = this.export();
		ret['account'] = encode(JSON.stringify(this.__data.account));
		ret['storage'] = encode(JSON.stringify(storage));
		ret['data'] = JSON.stringify(data);
		return ret;
	};
	exportSigned() {
		var ret = this.exportPacked();
		ret['dataSign'] = this.sign(ret['data']);
		ret['accSign'] = this.sign(ret['account']);
		return ret;
	};
}


export default Session;
