// preData/server.es6.js


import Logger from 'logger';

var logger = Logger.getLogger('preData');



var Storage = require('./preDataStorage.js');


var url = require('url');
var pathToRegexp = require('path-to-regexp');


String.prototype.hashCode = function() {
	var hash = 0, char;
	try {
		if (this.length == 0) return hash;
		for (var i = 0; i < this.length; i++) {
			char = this.charCodeAt(i);
			hash = ((hash << 5) - hash) + char;
			hash = hash & hash; // Convert to 32bit integer
		}
		return hash;
	} catch (e) {
		throw new Error('hashCode: ' + e);
	}
};



var Request = function(path, source, session = {}, body = {}, files = {}) {
	var data = url.parse(source);
	var keys = [];
	var query = {};
	var id = source.hashCode();
	!data.query?0:data.query.split('&').forEach(function(e) {
		try {
			var data = e.split('=');
			if(e.length > 0) {
				query[data[0]] = !data[1] ? null : (!isNaN(parseFloat(data[1])) && isFinite(data[1]) ? parseFloat(data[1]) : decodeURIComponent(data[1]));
			}
		} catch(e) {}
	});
	var re = pathToRegexp(path, keys);
	var params = {};
	var tmp1 = re.exec(data.pathname);
	if(data.pathname !== '\/' && data.pathname !== '' && tmp1) {
		var preParams = re.exec(data.pathname).map(function(e) {return e;})
		for(var i=1;i<preParams.length;i++) {
			var key = keys[i-1].name, val = preParams[i];
			params[key]=val;
		}
	}
	var req = {
		id: id,
		query: query,
		params: params,
		session: session,
		body: body,
		files: files,
		path: data.pathname,
		fullPath: data.path,
		queryString: data.query||''
	}
	return req;
};


class Res {
	constructor(res, next, title) {
		this.__res = res;
		this.__next = next;
		this.title = title;
		this.__data = {};
	}
	setData(name, data) {
		this.__data[name] = data;
		return this;
	}
	end(name) {
		this.__res.___res = this;
		this.__next();
	}
	redirect(url) {
		this.__res.redirect(url);
	}
	get data() {
		return this.__data;
	}
}

var Response = function(res, next, title) {
	return new Res(res, next, title);
};

var selectRoute = function(path, routes) {
	path = url.parse(path).pathname;
	return routes.filter(function(e) {
		var re = pathToRegexp(e.path);
		return !!re.exec(path);
	})[0]||null;
};


class PreData {
	constructor(app, routes) {
		app.post('/preData', (request, response, next)=> {
			var url = request.body.url;
			var body = {};
			var files = {};
			try {
				body = JSON.parse(request.body.body);
			} catch(e) {
				logger.error(e);
			}
			try {
				files = request.files;
			} catch(e) {
				logger.error(e);
			}
			if(request.___ret) return next();
			var route = selectRoute(url, routes);
			if(route) {
				var res, req;
				var _next = ()=> {
					response.json({
						preData: res.data,
						id: req.id
					});
				};
				req = Request(route.path, url, request.session, body, files);
				res = Response(response, _next, route.title);
				request.__ret = {
					req,
					res,
					status: route.status
				};
				route.express(req, res);
			} else {
				next();
			}
		});
		app.use('/preData', (request, response, next)=> {
			var url = request.query.url;
			var body = {};
			try {
				body = JSON.parse(request.query.body);
			} catch(e) {
				logger.error(e);
			}
			if(request.___ret) return next();
			var route = selectRoute(url, routes);
			if(route) {
				var res, req;
				var _next = ()=> {
					response.json({
						preData: res.data,
						id: req.id
					});
				};
				req = Request(route.path, url, request.session, body);
				res = Response(response, _next, route.title);
				request.__ret = {
					req,
					res,
					status: route.status
				};
				route.express(req, res);
			} else {
				next();
			}
		});
		app.use((request, response, next)=> {
			if(request.___ret) return next();
			var route = selectRoute(request.originalUrl, routes);
			if(route) {
				var req = Request(route.path, request.originalUrl, request.session, request.body);
				var res = Response(response, next, route.title);
				request.__ret = {
					req,
					res,
					status: route.status
				};
				route.express(req, res);
			} else {
				next();
			}
		});
	}
	create(data = {}) {
		let id = parseInt(Math.random().toString().substr(2)).toString(32);
		Storage[id] = data;
		return id;
	}
	get(id) {
		let data = Storage[id]||{};
		return data;
	}
	remove(id) {
		let data = this.get(id);
		delete Storage[id];
		return data;
	}
}







module.exports = PreData;
