// configs/index.js

var types = [
	"credentials",
	"api",
	"server",
	"application",
	"plugins",
	"routes",
	"style"
];

module.exports = function(type) {
	switch(type) {
		case "c":
		case "crd":
		case "crdnt":
		case "crdnts":
		case "crdntl":
		case "crdntls":
		case "crds":
			type = "credentials";
		break;
		case "srv":
		case "serv":
			type = "server";
		break;
		case "app":
			type = "application";
		break;
		case "p":
		case "plg":
		case "plgs":
			type = "plugins";
		break;
		case "r":
		case "rt":
		case "rts":
			type = "routes";
		break;
		case "st":
		case "stl":
		case "stls":
			type = "style";
		break;
	}
	if(types.indexOf(type) != -1) {
		if(type == 'credentials' || type == 'server') {
			if(typeof window != 'undefined') {
				return null;
			} else {
				var sources = {
					credentials: require('./../../configs/credentials'+'.json'),
					server: require('./../../configs/server'+'.json')
				};
				return sources[type];
			}
		} else {
			var sources = {
				api: require('./../../configs/api.json'),
				application: require('./../../configs/application.json'),
				plugins: require('./../../configs/plugins.json'),
				routes: require('./../../configs/routes.json'),
				style: require('./../../configs/style.json')
			};
			return sources[type];
		}
	} else {
		return null;
	}
}
